runtime/parser2/expression.go

```diff
@@ -876,3 +951,63 @@ func parseHex(r rune) rune {
 
 	return -1
 }
+
+func parseIntegerLiteral(p *parser, literal, text string, kind IntegerLiteralKind, tokenRange ast.Range) *ast.IntegerExpression {
+
+	report := func(invalidKind InvalidNumberLiteralKind) {
+		p.report(
+			&InvalidIntegerLiteralError{
+				IntegerLiteralKind:        kind,
+				InvalidIntegerLiteralKind: invalidKind,
+				// NOTE: not using text, because it has the base-prefix stripped
+				Literal: literal,
+				Range:   tokenRange,
+			},
+		)
+	}
+
+	// check literal has no leading underscore
+
+	if strings.HasPrefix(text, "_") {
+		report(InvalidNumberLiteralKindLeadingUnderscore)
+	}
+
+	// check literal has no trailing underscore
+	if strings.HasSuffix(text, "_") {
+		report(InvalidNumberLiteralKindTrailingUnderscore)
+	}
+
+	withoutUnderscores := strings.Replace(text, "_", "", -1)
+
+	base := kind.Base()
+	value, ok := new(big.Int).SetString(withoutUnderscores, base)
+	if !ok {
+		report(InvalidNumberLiteralKindUnknown)
+	}
+
+	return &ast.IntegerExpression{
+		Value: value,
+		Base:  base,
+		Range: tokenRange,
+	}
+}
+
+func parseFixedPointPart(part string) (integer *big.Int, scale uint) {
+	withoutUnderscores := strings.Replace(part, "_", "", -1)
+	integer, _ = new(big.Int).SetString(withoutUnderscores, 10)
+	return integer, uint(len(withoutUnderscores))
+}
+
+func parseFixedPointLiteral(text string, tokenRange ast.Range) *ast.FixedPointExpression {
+	parts := strings.Split(text, ".")
+	integer, _ := parseFixedPointPart(parts[0])
+	fractional, scale := parseFixedPointPart(parts[1])
+
+	return &ast.FixedPointExpression{
+		Negative:        false,
+		UnsignedInteger: integer,
```

Comment:
Just curious why we manage the signed-ness ourselves versus using a `Integer` field in place of `UnsignedInteger` and `Negative` that accepts signed ints.

Reply:
The code initially did that, i.e. it only had an `Integer` field instead of the two `Negative` and `UnsignedInteger` flags.

However, `big.Int` doesn't support negative zeros, so representing e.g. `-0.1`, which has the integer part `-0` wasn't possible.